### ISSUE TITLE
[Feature]: Make token_ids a first-class output type and enforce canonical names

### ISSUE THREAD

**Issue by Jungle430** [vllm-project/vllm-omni#7841]
## Current behavior (before this change)

[`output_modality.py`](https://github.com/vllm-project/vllm-omni/blob/23d8c83dab27037332205b376d4986dec158b366/vllm_omni/outputs/output_modality.py#L13) currently accepts aliases:

```python
_MODALITY_ALIASES: dict[str, str] = {
    "speech": "audio",
    "images": "image",
    "latents": "latent",
    "wav": "audio",
    "waveform": "audio",
    "pixel_values": "image",
    "pixels": "image",
    "token_ids": "text",
    "tokens": "text",
}
```

`OutputModality.from_string()` (docstring omitted):

```python
@classmethod
def from_string(cls, s: str | None) -> OutputModality:
    if not s or not s.strip():
        return cls.TEXT

    parts = [p.strip().lower() for p in re.split(r"[+,]", s.strip())]
    result = cls(0)
    for p in parts:
        p = _MODALITY_ALIASES.get(p, p)
        try:
            result |= cls[p.upper()]
        except KeyError:
            raise ValueError(f"Unknown modality: {p!r}. Supported: {[m.name.lower() for m in cls]}")
    return result
```

### Why the output type matters

`PipelineConfig → resolved stage config → runner / output processor`. Runners can inspect the raw string; output processing converts it to an enum and derives a default `mm_type`, which affects:

1. **Output fields.** [`MultimodalPayload.from_raw()`](https://github.com/vllm-project/vllm-omni/blob/23d8c83dab27037332205b376d4986dec158b366/vllm_omni/outputs/mm_outputs.py#L245) uses it to name generic producer payloads:

   ```python
   # Same producer payload, different default output keys:
   MultimodalPayload.from_raw({"model_outputs": tensor}, "audio").to_dict()
   # {"audio": tensor}  (moved to CPU)
   MultimodalPayload.from_raw({"model_outputs": tensor}, "latent").to_dict()
   # {"latent": tensor}  (moved to CPU)
   ```

   Bare tensors and generic `hidden` fields use the same naming rule; explicit keys such as `ids.prior_image` remain unchanged.

2. **Tensor merging.** [`get_accumulation_strategy()`](https://github.com/vllm-project/vllm-omni/blob/23d8c83dab27037332205b376d4986dec158b366/vllm_omni/outputs/output_modality.py#L143) selects the strategy, and [`_cat_tensors()`](https://github.com/vllm-project/vllm-omni/blob/23d8c83dab27037332205b376d4986dec158b366/vllm_omni/outputs/mm_outputs.py#L35) performs the merge. The default behavior is equivalent to:

   ```python
   chunks = [torch.zeros(1, 3), torch.ones(1, 3)]
   if OutputModality.AUDIO in modality:
       merged = torch.cat(chunks, dim=-1)  # shape: (1, 6)
   else:  # TEXT / IMAGE / LATENT
       merged = torch.cat(chunks, dim=0)   # shape: (2, 3)
   ```

   Registered per-key strategies take precedence over these defaults.

Reject noncanonical values during service initialization, before accepting requests, and represent discrete token IDs explicitly.

## Proposed changes

- Remove `_MODALITY_ALIASES`; accept canonical `text`, `image`, `audio`, `latent`, and `token_ids`.
- Make `token_ids` an independent type for discrete vocabulary/codebook indices, rather than a `text` alias. Distinguish it from continuous latent representations without migrating existing models that use `latent` more broadly.
- Preserve `None` defaults and `+` / `,` combinations. Reject other aliases, unknown names, mixed case, whitespace, empty strings, and empty components. External configs and direct callers must migrate explicitly.
- Preserve explicit empty values through config projection. Let `from_string()` reject invalid values when the LLM output processor is constructed during initialization. Replicas may already have started; initialization failure cleans up acquired clients. This does not promise validation before any engine launch.
- Wire the new type through default output naming and non-drainable DELTA handling. Keep existing dim-0 tensor accumulation and per-key overrides. Generic `model_outputs` and `hidden` follow the shared default-key naming rule, without a `token_ids`-specific exception.

## GLM-Image

Keep the AR stage's original declaration:

```python
engine_output_type="token_ids"
```

The new parsing result is `OutputModality.TOKEN_IDS`, not `TEXT` or `LATENT`. The proposed migration to `latent` is withdrawn following the discussion below.

Generated IDs still travel through `cumulative_token_ids`; source-image IDs remain in `ids.prior_image`. `ar2diffusion()` continues reading these fields. When present, auxiliary `hidden` is normalized to `token_ids` without changing its tensor values. GLM-Image’s bridge ignores this auxiliary field and reads generated IDs from `cumulative_token_ids`. The default GPU path omits auxiliary hidden output; NPU and some optional runner/cache paths may still emit it and require separate validation.

No changes to `final_output_type`, Diffusion's `output_type`, model execution, or tensor-concatenation failure policies. Existing `latent` pipelines remain unchanged. A richer per-field contract separating modality, representation, and update/retention rules is follow-up work.

## Validation

- [ ] Cover canonical/invalid values, combinations, `None`, empty-value preservation, and startup failure/cleanup.
- [ ] Cover token-ID routing, shared field normalization, accumulation and DELTA retention; verify that explicitly named auxiliary fields remain intact and built-in output declarations are canonical.
- [ ] Verify GLM T2I/I2I bridge tokens and source-image conditioning, then compare real NVIDIA runs against the base. NPU coverage remains separate.

cc @tzhouam

**Comment by tzhouam** [comment]
I agree with validating canonical values before engine startup, but I have a concern about migrating GLM-Image stage 0 to `latent`.

The AR stage produces discrete visual token IDs associated with the VQ-VAE codebook. These are distinct from the continuous diffusion/VAE latents that are denoised and decoded into pixels. The [Transformers implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/glm_image/modeling_glm_image.py) explicitly describes the image tokens as discrete VQ-VAE codebook indices, while the [Diffusers pipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/glm_image/pipeline_glm_image.py) handles AR image tokens and continuous latents separately.

Discrete codes can be considered latent representations in a broad sense, but we should define whether `engine_output_type="latent"` intentionally covers both representations. Avoiding the runner's raw `"text"` filtering and sharing accumulation behavior establishes operational compatibility, but does not by itself establish the right type semantics.

Could we separate canonical-value validation from this migration, and consider making `token_ids` a first-class type (rather than a `text` alias), or introducing an explicit `image_tokens` type? If we retain `latent`, its contract should explicitly distinguish discrete token IDs from continuous latent tensors so downstream consumers cannot assume the latter.

**Comment by Jungle430** [comment]
Thanks for pointing out this distinction. I agree that compatible output-processing behavior alone does not justify classifying GLM-Image’s discrete token IDs as `latent`.

Semantically, I would distinguish:

- **`token_ids`**: discrete symbol indices into a vocabulary or codebook, such as the image token IDs produced by GLM-Image’s AR stage.
- **`latent`**: continuous latent representations used for further computation, denoising, or decoding.

Both can be intermediate outputs, but downstream consumers interpret them differently. This distinction concerns their meaning, not simply whether a tensor has an integer or floating-point dtype.

For this PR, I plan to retain canonical-value validation and make `token_ids` a first-class type rather than a `text` alias. GLM-Image stage 0 would keep `engine_output_type="token_ids"`, reverting the migration to `latent`, with corresponding output-handling changes and tests. Existing `latent` usage in other models would remain unchanged; the distinction above is the intended direction, not a retroactive restriction on all existing uses in this PR.

For a follow-up design, I would separate the descriptive and processing responsibilities currently carried by a single `engine_output_type`:

1. **Describe modality and representation per output field.** For example, `image + token_ids`, `image + latent`, and `audio + waveform`, with explicit representation types for hidden states and embeddings.
2. **Allow multiple named output fields per stage.** A stage could produce generated image IDs, reference image IDs, and hidden states without assigning them one shared type label. Whether an output is returned to the user would remain a separate decision.
3. **Declare update and retention rules per field.** Incremental chunks could concatenate along a specified axis, cumulative snapshots could replace previous values, and reference data could remain available for the request. Codebook and layout requirements could be specified where needed, rather than inferring processing behavior from modality alone.

This could build on the existing payload fields and gradually establish a clearer output contract. This PR would focus on making `token_ids` explicit, leaving the broader design for a follow-up discussion.

**Comment by tzhouam** [comment]
Looks good, maybe we can start with this and have a more robust and  clear interface for the output modalities later.

**Comment by Jungle430** [comment]
Thanks for confirming! I’ve updated #7843 to make `token_ids` a first-class output type for discrete vocabulary/codebook indices. GLM-Image keeps `engine_output_type="token_ids"`, and existing models using `latent` are unchanged. Canonical-name validation, the corresponding output handling, and tests are included.

The 56 self-contained tests pass. The full related test suite and NVIDIA GLM-Image T2I/I2I regression are still pending, so the PR remains a draft. The richer per-field contract for modality, representation, and update rules will be follow-up work.
